src/openstack_manager.py

```diff
@@ -113,3 +85,196 @@ def list_projects(cloud_config: dict) -> list[Project]:
         ) from exc
 
     return projects
+
+
+class ImageBuildError(Exception):
+    """Exception representing an error during image build process."""
+
+
+def _build_image_command(
+    runner_info: RunnerApplication, proxies: Optional[ProxyConfig] = None
+) -> list[str]:
+    """Get command for building runner image.
+
+    Returns:
+        Command to execute to build runner image.
+    """
+    http_proxy = proxies.http if (proxies and proxies.http) else ""
+    https_proxy = proxies.https if (proxies and proxies.https) else ""
+    no_proxy = proxies.no_proxy if (proxies and proxies.no_proxy) else ""
+
+    environment = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+    docker_proxy_service_conf_content = environment.get_template("systemd-docker-proxy.j2").render(
+        proxies=proxies
+    )
+    docker_proxy = {
+        "proxies": {
+            "default": {
+                key: value
+                for key, value in (
+                    ("httpProxy", http_proxy),
+                    ("httpsProxy", https_proxy),
+                    ("noProxy", no_proxy),
+                )
+                if value
+            }
+        }
+    }
+    docker_client_proxy_content = json.dumps(docker_proxy)
+
+    cmd = [
+        "/usr/bin/bash",
+        BUILD_OPENSTACK_IMAGE_SCRIPT_FILENAME,
+        runner_info["download_url"],
+        str(http_proxy),
+        str(https_proxy),
+        str(no_proxy),
+        docker_proxy_service_conf_content,
+        docker_client_proxy_content,
+    ]
+
+    return cmd
+
+
+@dataclass
+class InstanceConfig:
+    """The configuration values for creating a single runner instance.
+
+    Args:
+        name: Name of the image to launch the GitHub runner instance with.
+        labels: The runner instance labels.
+        registration_token: Token for registering the runner on GitHub.
+        github_path: The GitHub repo/org path
+        openstack_image: The Openstack image to use to boot the instance with.
+    """
+
+    name: str
+    labels: Iterable[str]
+    registration_token: str
+    github_path: GithubPath
+    openstack_image: openstack.image.v2.image.Image
```

Comment:
The interfaces are subject to change during refactor - which is out of the scope of this PR.